src/workerd/api/filesystem.h

```diff
@@ -0,0 +1,333 @@
+#pragma once
+
+#include <workerd/api/streams/writable.h>
+#include <workerd/io/worker-fs.h>
+#include <workerd/jsg/iterator.h>
+#include <workerd/jsg/jsg.h>
+
+namespace workerd::api {
+
+class Blob;
+class File;
+
+// ======================================================================================
+// An implementation of the WHATWG Web File System API (https://fs.spec.whatwg.org/)
+// All of the classes in this part of the impl are defined by the spec.
+
+class FileSystemSyncAccessHandle;
+class FileSystemWritableFileStream;
+
+class FileSystemHandle: public jsg::Object {
+  // TODO(node-fs): The spec defines FileSystemHandle objects as being
+  // serializable, meaning that they should work with structured cloning.
+  // In workers, this would suggest also that they should work with jsrpc.
+  // We don't yet implement any form of serialization for these objects.
+  // Since each worker has its own file system, it might be a bit weird
+  // as the received file on the other side would not actually be in that
+  // destination worker's file system... that is, it would be an orphaned
+  // file handle. It would still likely be useful tho. As a follow up
+  // step, we will implement serialization/deserialization of these objects.
+ public:
+  FileSystemHandle(jsg::USVString name): name(kj::mv(name)) {}
+  jsg::USVString getName(jsg::Lock& js) {
+    return jsg::USVString(kj::str(name));
```

Comment:
Why do we need the double conversion?
Isn't it enough to just return name?

Reply:
`jsg::USVString` does not have an equivalent to `kj::StringPtr` and the type mapping requires `jsg::USVString` for it to get written correctly. It's an annoyance at this point that I plan to get back around to simplifying but not in this PR.